overlord/storestate/storestate.go

```diff
@@ -0,0 +1,165 @@
+	API string `json:"api"`
+}
+
+// API returns the store's explicit address.
+func API(st *state.State) string {
+	var storeState StoreState
+
+	err := st.Get("store", &storeState)
+	if err != nil {
+		return ""
+	}
+
+	return storeState.API
+}
+
+// updateAPI writes the store's address to persistent state.
+func updateAPI(st *state.State, api string) {
+	var storeState StoreState
+	st.Get("store", &storeState)
+	storeState.API = api
+	st.Set("store", &storeState)
+}
+
+// A StoreService can find, list available updates and download snaps.
+type StoreService interface {
+	SnapInfo(spec store.SnapSpec, user *auth.UserState) (*snap.Info, error)
+	Find(search *store.Search, user *auth.UserState) ([]*snap.Info, error)
+	LookupRefresh(*store.RefreshCandidate, *auth.UserState) (*snap.Info, error)
+	ListRefresh([]*store.RefreshCandidate, *auth.UserState) ([]*snap.Info, error)
+	Sections(user *auth.UserState) ([]string, error)
+	Download(context.Context, string, string, *snap.DownloadInfo, progress.Meter, *auth.UserState) error
+
+	Assertion(assertType *asserts.AssertionType, primaryKey []string, user *auth.UserState) (asserts.Assertion, error)
+
+	SuggestedCurrency() string
+	Buy(options *store.BuyOptions, user *auth.UserState) (*store.BuyResult, error)
+	ReadyToBuy(*auth.UserState) error
+}
+
+// SetupStore configures the system's initial store.
+func SetupStore(st *state.State, authContext auth.AuthContext) error {
+	storeConfig, err := initialStoreConfig(st)
+	if err != nil {
+		return err
+	}
+	sto := storeNew(storeConfig, authContext)
+	replaceAuthContext(st, authContext)
+	ReplaceStore(st, sto)
+	return nil
+}
+
+// SetupStoreAPI replaces the API of the store used by the system. If the API
+// URL is nil the store is reverted to the system's default.
+func SetupStoreAPI(state *state.State, api *url.URL) error {
+	apiState := ""
+	config := store.DefaultConfig()
+	if api != nil {
+		apiState = api.String()
+		err := config.SetAPI(api)
+		if err != nil {
+			return err
+		}
+	}
+	authContext := authContext(state)
+	store := store.New(config, authContext)
+	ReplaceStore(state, store)
+	updateAPI(state, apiState)
+	return nil
+}
+
+func initialStoreConfig(st *state.State) (*store.Config, error) {
+	config := store.DefaultConfig()
+	apiState := API(st)
+	if apiState != "" {
+		api, err := url.Parse(apiState)
+		if err != nil {
+			return nil, fmt.Errorf("invalid store API URL: %s", err)
+		}
+		err = config.SetAPI(api)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return config, nil
+}
+
+type cachedAuthContextKey struct{}
+
+func replaceAuthContext(state *state.State, authContext auth.AuthContext) {
+	state.Cache(cachedAuthContextKey{}, authContext)
+}
+
+func authContext(state *state.State) auth.AuthContext {
+	cached := state.Cached(cachedAuthContextKey{})
+	if cached != nil {
+		return cached.(auth.AuthContext)
+	}
+	panic("internal error: needing the auth context before managers have initialized it")
+}
```

Comment:
I still wonder if it wouldn't be simpler to give store.Store and store.StoreService an accessor to get the current AuthContext, it comes from outside anyway, so really no big harm afaics

Reply:
I'm open to changing this. However, I think hanging on to the auth context for later is a hack anyway, and I'd prefer to contain the hack to internal helpers where it's more likely to get cleaned up if the hack ever becomes unnecessary.

Reply:
ok